Refactor block volume's descriptor lock logic

mkimuram opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR is a refactoring of block volume's descriptor lock logic.

There are two issues in the current implementation:
1. Lock is taken to the device itself, so it is per volume. However, the same volume can be used by multiple pods and some drivers require the lock to be released per pod (before ensuring that there are no other pods using the volume),  
2. Lock logic is called inside each driver, which makes it difficult to restrict the order of handling lock logic.

This PR fix these issues with below changes:
1. Change lock to be taken to the file under globalMapPath, which is per pod,
2. Change globalMapPath to bind mount from symlink (losetup resolves the symlink, so we can't distinguish the path unless it is changed to bind mount), 
3. Move lock logic out from driver. 

This PR is needed for #73773 to separate staging/publish and unstaging/unpublish logics for CSI driver.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:
/sig storage

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

mkimuram committed:
Change globalMapPath to bind mount from symlink

This change is needed to make descriptor lock per pod, in the next commit.
If losetup is called for symlink, path in the output for losetup is resolved,
as a result, we can't distinguish which path the lock is taken.

mkimuram committed:
Make descriptor lock per pod and release it per pod

This change is needed to avoid unmapVolumeFunc for one pod blocked,
when the lock for the same volume is taken for another pod.

mkimuram committed:
Move MapBlockVolume call to operation_generator and add UnmapBlockVolume

mkimuram committed:
Check and return error first in IsSymlinkExist and IsBindMountExist

mkimuram committed:
Improve comments for volume path hanlder and volume.go

mkimuram committed:
Remove klog for output error instead return err with context

mkimuram commented:
@davidz627 @wongma7 

I think that I fixed other than below two comments in the last three commits. PTAL

> https://github.com/kubernetes/kubernetes/pull/74026#discussion_r340795799

It is doing almost the same logic to `IsSymlinkExist`, so it would be good enough.
(Also, calling methods like `IsNotMountPoint` won't return the right value for this particular case,
for it won't care if it is a bind mount of a device.)


> https://github.com/kubernetes/kubernetes/pull/74026#discussion_r340794511

It is creating mounter there just for handling bind mount/unmount in plugin independent code.
So, I think that we can use any mounter, which can handle bind mount properly, here for all plugins.
(Also, for block volume, no existing plugin creates mounter that we can reuse there.)

In this case, should we create a mounter outside volume_path_handler and pass mounter to it?  
(We might be able to do similar to [GenerateMountVolumeFunc](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L644) in `GenerateMapVolumeFunc` and `GenerateUnmapVolumeFunc`, and pass the mounter to voume_path_handler, though.)

mkimuram commented:
/retest

jsafrane commented:
/retest

jsafrane commented on pkg/volume/util/volumepathhandler/volume_path_handler.go:
```diff
@@ -23,28 +23,32 @@ import (
-	// GetDeviceSymlinkRefs searches symbolic links under global map path
-	GetDeviceSymlinkRefs(devPath string, mapPath string) ([]string, error)
+	// IsBindMountExist retruns true if specified bind mount exists
+	IsBindMountExist(mapPath string) (bool, error)
```
Please document that it checks only bind-mounted block devices. Or rename the function to IsDeviceBindMount or something like that. It is not a generic bind-mount check.

mkimuram replied:
Renamed to `IsDeviceBindMount` as suggested.

jsafrane commented on pkg/volume/util/volumepathhandler/volume_path_handler.go:
```diff
@@ -98,25 +102,105 @@ func (v VolumePathHandler) MapDevice(devicePath string, mapPath string, linkName
+			return fmt.Errorf("Failed to stat file %s: %v", linkPath, err)
+		}
+
+		klog.Warningf("Warning: Path to bind mount %v has not yet been created", linkPath)
```
Why `Warning`?  It is going to be logged for every bind-mount.

mkimuram replied:
Removed the warning.

jsafrane commented on pkg/volume/util/volumepathhandler/volume_path_handler.go:
```diff
@@ -98,25 +102,105 @@ func (v VolumePathHandler) MapDevice(devicePath string, mapPath string, linkName
+		return checkErr
+	} else if !isMountExist {
+		klog.Warningf("Warning: Unmap skipped because bind mount does not exist on the path: %v", linkPath)
+		// TODO: consider deleting empty file if it exists
```
yes, please remove the empty file.

mkimuram replied:
Added a logic to remove empty file.

jsafrane commented on pkg/volume/util/volumepathhandler/volume_path_handler_linux.go:
```diff
@@ -93,16 +103,97 @@ func (v VolumePathHandler) RemoveLoopDevice(device string) error {
+
+func getDeviceMajorMinor(path string) (string, error) {
+	args := []string{"-c", "%t:%T", path}
+	cmd := exec.Command(statPath, args...)
```
Please use `unix.Stat` instead of exec, if possible

mkimuram replied:
Changed to use unix.Stat.

jsafrane commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1084,21 +1085,19 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
 		if mapErr != nil {
 			// On failure, return error. Caller will log and retry.
-			return volumeToMount.GenerateError("MapVolume.MapDevice failed", mapErr)
+			return volumeToMount.GenerateError("MapVolume.MapPodDevice failed", mapErr)
```
Don't log non-existing function name, use something like `MapVolume.MapDevice failed for pod local device`.

mkimuram replied:
Fixed.

jsafrane commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1084,21 +1085,19 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
+		mapErr = ioutil.MapBlockVolume(og.blkUtil, devicePath, globalMapPath, volumeMapPath, volName, volumeToMount.Pod.UID)
+		if mapErr != nil {
+			// On failure, return error. Caller will log and retry.
+			return volumeToMount.GenerateError("MapVolume.MapDevice failed", mapErr)
```
`MapBlockVolume failed`?

mkimuram replied:
Fixed this and similar typos.

jsafrane commented on pkg/volume/util/volumepathhandler/volume_path_handler.go:
```diff
@@ -97,11 +97,10 @@ func (v VolumePathHandler) MapDevice(devicePath string, mapPath string, linkName
 	if err != nil && !os.IsNotExist(err) {
-		klog.Errorf("cannot validate map path: %s", mapPath)
-		return err
+		return fmt.Errorf("Cannot validate map path: %s: %v", mapPath, err)
```
errors start with lowercase (`"cannot validate..."`), so they can be easily chained with `":"`

Mockup of such chained error:
"pod startup failed: Cannot validate map path: permission denied".

This applies to whole PR.

mkimuram replied:
Fixed similar issues in this PR.

jsafrane reviewed (commented):
Checking just the code sanity, got lost in OperationGenerator.

mkimuram committed:
Rename IsBindMountExist to IsDeviceBindMountExist

mkimuram committed:
Remove remaining empty file in unmapBindMountDevice

mkimuram committed:
Fix error messages in operation_generator.go

mkimuram committed:
Change getDeviceMajorMinor to use unix.Stat

mkimuram commented:
/retest

mkimuram commented:
@jsafrane 

Fixed as suggested and all tests passed. PTAL

msau42 commented:
/assign @saad-ali

jsafrane commented:
/priority important-longterm
/lgtm
/approve

mkimuram committed:
Change mount.NewOSExec to utilexec.New

mkimuram commented:
@jsafrane 

I noticed 367f879#diff-1109319b71777d5f4d35ba198db708e7 also affect this PR.
Rebased with the latest master and fixed manually in the last commit. PTAL and lgtm again

jsafrane commented:
/lgtm
/retest

saad-ali commented:
/milestone v1.17

saad-ali commented on pkg/volume/awsebs/aws_ebs_block.go:
```diff
@@ -148,7 +147,7 @@ func (b *awsElasticBlockStoreMapper) SetUpDevice() (string, error) {
 }
 
 func (b *awsElasticBlockStoreMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, volumeMapName string, podUID types.UID) error {
```
Should `MapDevice(` be removed from the interface if it is not used anywhere?

saad-ali replied:
Oh I see it is handled in https://github.com/kubernetes/kubernetes/pull/84747

saad-ali reviewed (approved):
/approve
